### PR TITLE
Fix the Adam step counter saturating under a BF16 default dtype

### PR DESCRIPTION
**Pull request by RachelWanggg**
`_AdamBase.step()` initializes the step counter as `torch.tensor(0.0)`, which picks up
`torch.get_default_dtype()` instead of being pinned to FP32. Under
`torch.set_default_dtype(torch.bfloat16)` the counter is a BF16 tensor, and BF16 cannot
represent 257, so `state["step"] += 1` stops advancing at 256.

`bias_correction1` / `bias_correction2` are derived from that frozen step, so
`bias_correction2` stays pinned at `1 - 0.999**256 = 0.226` instead of converging to 1.
`denom = sqrt(v) / sqrt(bias_correction2)` is then permanently ~2.1x too large, and every
update after step 256 is silently too small. There is no warning — training just gets a
quietly wrong effective learning rate.

### Repro

Constant gradient, `lr=1e-3`, 1000 steps, single FP32 parameter initialized to 1.0:

```python
import torch
from torchao.optim import _AdamW

def run(default_dtype, steps=1000):
    torch.set_default_dtype(default_dtype)
    p = torch.nn.Parameter(torch.ones(4, dtype=torch.float32))
    opt = _AdamW([p], lr=1e-3, weight_decay=0.0)
    for _ in range(steps):
        p.grad = torch.full_like(p, 0.1)
        opt.step()
        opt.zero_grad()
    step = opt.state[p]["step"]
    torch.set_default_dtype(torch.float32)
    return step.item(), step.dtype, p[0].item()
```

Before this change:

```
torchao _AdamW    default=fp32 : step=1000.0 (torch.float32)   param=0.00001004
torchao _AdamW    default=bf16 : step= 256.0 (torch.bfloat16)  param=0.20410579
torch.optim.AdamW default=bf16 : step=1000.0 (torch.float32)   param=0.00001008
```

After: all three agree at `step=1000`, `param≈1.00e-5`.

The step tensor lives in shared `_AdamBase` code, so `Adam8bit`, `AdamW8bit`, `Adam4bit`,
`AdamW4bit`, `AdamFp8` and `AdamWFp8` are all affected.

### Why an integer counter

A float counter stops advancing once `+1` falls below one ULP. BF16 hits that at 256, and
FP32 hits it at `2**24 = 16,777,216` — so pinning to FP32 would move the bug rather than
remove it. int32 is exact to `2**31-1`.

`step` is only consumed as `beta1**step` / `beta2**step`, and that returns the same FP32
result bit-for-bit whether `step` is int32 or FP32 (checked across the four common beta
values for every step count from 1 to `2**24-1`), so nothing downstream changes.
`_AdamBase.add_param_group` already pins `lr` to FP32 explicitly, so pinning the step dtype
matches the existing convention in this file.

Checkpoints written before this change are unaffected: a saved FP32 step loads back as FP32
and keeps counting.

BF16 as the default dtype is a supported configuration here — `torchao/optim/README.md`
documents full-BF16 training and describes these optimizers as a drop-in replacement for
`torch.optim.Adam`, which stays correct in the same configuration via
`torch.optim.optimizer._get_scalar_dtype()`.

### Test

`test_optim_default_dtype_bf16` already sets `torch.set_default_dtype(torch.bfloat16)`, but
runs a single `optimizer.step()` and asserts nothing about the result, so it is far below
the 256-step saturation point. Added `test_optim_default_dtype_bf16_step_counter`, which
asserts the step counter is not a floating-point dtype and that it advances past 256.

`pytest test/test_low_bit_optim.py` shows no new failures. `ruff check` and `ruff format`
(0.11.6, per CONTRIBUTING) are clean.

### CODE REVIEW

**Comment by gau-nernst** [comment]
i wonder if we should use int32. torch compile should fuse the dtype conversion anyway

**Comment by RachelWanggg** [comment]
Thank you, good call, switched to int32. FP32 caps out at 16M steps anyway, so it'd just move the problem. Tested `beta**step` (bit-identical) and fullgraph compile passes. Backward compatible with old checkpoints too.

**Review by gau-nernst — APPROVED** [review]
